File: web/root.go
// Townsourced
// Copyright 2015 Tim Shannon. All rights reserved.

package web

import (
	"net/http"
	"time"

	log "git.townsourced.com/townsourced/logrus"
	"git.townsourced.com/townsourced/townsourced/app"
	"git.townsourced.com/townsourced/townsourced/data"
)

const postPageLimit = 40

// root page handler
func rootTemplate(w http.ResponseWriter, r *http.Request, c context) {
	if demoMode {
		err := w.(*templateWriter).execute("DEMO", nil)
		if err != nil {
			log.Errorf("Error executing DEMO template: %s", err)
		}
		return
	}

	if c.session == nil {
		location, err := app.IPToLocation(ipAddress(r))
		if errHandledPage(err, w, r, c) {
			return
		}

		local, towns, err := app.PostGetByLocation(location.Longitude, location.Latitude, 300, "", time.Time{}, 6)
		if errHandledPage(err, w, r, c) {
			return
		}

		if len(local) < 6 {
			towns, err = app.IPToTowns(ipAddress(r), 6)
			if errHandledPage(err, w, r, c) {
				return
			}
		}

		err = w.(*templateWriter).execute("PUBLIC", struct {
			Local []app.Post
			Towns []app.Town
		}{
			Local: local,
			Towns: towns,
		})
		if err != nil {
			log.Errorf("Error executing public template: %s", err)
		}
		return
	}

	u, err := c.session.User()

	if errHandledPage(err, w, r, c) {
		return
	}

	posts, err := app.PostGetByTowns(u, data.KeyWhenSlice(u.TownKeys).Keys(), "", time.Time{}, postPageLimit, false)
	if errHandledPage(err, w, r, c) {
		return
	}

	var towns []app.Town
	townHint := false

	if len(posts) < postPageLimit || len(u.TownKeys) < 2 {
		townHint = true
		localTowns, err := app.IPToTowns(ipAddress(r), 10)
		if errHandledPage(err, w, r, c) {
			return
		}

		// add only towns that they aren't already a member of
		for i := range localTowns {
			found := false
			for j := range u.TownKeys {
				if u.TownKeys[j].Key == localTowns[i].Key {
					found = true
				}
			}
			if !found {
				towns = append(towns, localTowns[i])
			}
		}
	}

	err = w.(*templateWriter).execute("ROOT", struct {
		Posts    []app.Post
		User     *app.User
		TownHint bool
		Towns    []app.Town
	}{
		Posts:    posts,
		User:     u,
		TownHint: townHint,
		Towns:    towns,
	})

	if err != nil {
		log.Errorf("Error executing ROOT template: %s", err)
	}
}

File: web/src/js/root.js
// Townsourced
// Copyright 2015 Tim Shannon. All rights reserved.
/* jshint  esnext: true, strict: true */

// Ractive + components
import Ractive from "./ts/ractiveInit";
import Page from "./components/page";
import PostList from "./components/postList";
import TownList from "./components/townList";
import Header from "./components/header";
import Alert from "./components/alert";
import CategorySelect from "./components/categorySelect";
import TownPanel from "./components/townPanel";
import Modal from "./components/modal";
import PostOptions from "./components/postOptions";


// ts libs
import {
    setTheme,
    join,
    leave,
}
from "./ts/town";
import {
    htmlPayload,
    pluck,
    scrollToFixed,
}
from "./ts/util";
import {
    buildSearchParams,
}
from "./ts/search";

import {
    towns as getUserTowns,
}
from "./ts/user";

import {
    err
}
from "./ts/error";


//3rd party


$(document).ready(function() {
    "use strict";

    var r = new Ractive({
        el: "body",
        template: "#tMain",
        components: {
            page: Page,
            postList: PostList,
            alert: Alert,
            townList: TownList,
            header: Header,
            categorySelect: CategorySelect,
            modal: Modal,
            townPanel: TownPanel,
            postOptions: PostOptions,
        },
        data: function() {
            var towns = htmlPayload("townsPayload") || [];
            return {
                searchOptions: {
                    category: "all",
                    towns: [],
                },
                posts: htmlPayload("postsPayload"),
                currentUser: htmlPayload("userPayload"),
                towns: towns.slice(0, 3),
                town: null,
                error: null,
                userTowns: [],
            };
        },
    });

    var postList = r.findComponent("postList");

    scrollToFixed(r.nodes.navCenter, r.nodes.fixedTrigger);
    scrollToFixed(null, r.nodes.fixedTrigger, null, function(fixed) {
        r.set("fixed", fixed);
    });


    r.on({
        "navbar.userLoaded": function(currentUser) {
            if (currentUser) {
                userTowns();
                return;
            }
        },
        "search": function(event) {
            event.original.preventDefault();

            if (!r.get("searchOptions.search").trim()) {
                return;
            }
            window.location = "/search?" + buildSearchParams({
                search: r.get("searchOptions.search"),
                category: r.get("searchOptions.category"),
            });
        },
        "changeCategory townList.select townList.deselect": function(event) {
            postList.fire("getPosts");
        },
        "selectTown": function(event) {
            event.original.preventDefault();
            r.set("town", event.context);

            setTheme(r.get("town.color"));
            $("#infoPanel").modal();
        },
        "townPanel.join": function(event) {
            event.original.preventDefault();
            join(r.get("town.key"))
                .done(function(result) {
                    location.reload();
                })
                .fail(function(result) {
                    r.set("townError", err(result).message);
                });
        },
        "townPanel.leave": function(event) {
            event.original.preventDefault();
            leave(r.get("town.key"))
                .done(function(result) {
                    location.reload();
                })
                .fail(function(result) {
                    r.set("townError", err(result).message);
                });
        },
        "post": function(event) {
            event.original.preventDefault();
            r.findComponent("postOptions").fire("show");
        },
    });

    function userTowns() {
        getUserTowns()
            .done(function(result) {
                r.set("userTowns", result.data);
            })
            .fail(function(result) {
                r.set("error", err(result).message);
            });
    }


});
